### PR TITLE
Require input and output amounts for raw interactions

### PR DESCRIPTION
**Pull request by nlordell**
This PR augments interaction data to also include input and output amounts. This will be required by the driver in order to verify token conservation across tokens and orders.

### Test Plan

Added unit test to verify de-serialization. For now, there is no new logic changes.

### Release notes

This PR introduces changes to the HTTP solver interface. Note that the API was kept backwards compatible to allow solver implementations time to adapt to the new API. These changes will be made required after a certain amount of time (since they are required for validating new solver rules around local and global token conservation).

### CODE REVIEW

**Review by MartinquaXD — COMMENTED** [review]
The code itself looks good (nice refactor of `TokenAmount`).
> verify token conservation across tokens and orders...

Could harm be done by solvers lying about the input and output amounts?
I guess we would have to simulate each interaction to know whether those values are correct, right?

**Comment by nlordell** [comment]
> Could harm be done by solvers lying about the input and output amounts?

Yes. What is unclear is whether or not we should verify this upfront (by simulating as you suggested) or if we should slash solvers that misbehave.

File: crates/shared/src/http_solver/model.rs [hunk 1]
```diff
@@ -121,11 +115,15 @@ pub struct ApprovalModel {
...
 
-#[derive(Clone, Debug, Deserialize, PartialEq)]
+#[derive(Clone, Debug, Derivative, Deserialize, PartialEq)]
 pub struct InteractionData {
     pub target: H160,
     pub value: U256,
+    #[derivative(Debug(format_with = "debug_bytes"))]
+    #[serde(with = "model::bytes_hex")]
     pub call_data: Vec<u8>,
+    pub input: TokenAmount,
```
[the remark this answers is not in the document]
Comment (nlordell) [reply crates/shared/src/http_solver/model.rs:125]: Makes sense! I forget that some of these terms are loaded with context and my assumption != your assumption.

I will add a comment.

**Review by vkgnosis — COMMENTED** [review]
I guess this has been discussed already but doesn't this limit what an interaction can do in a potential important manner? What if you have an interaction with one input amount and two output amounts or vice versa? What if a first interaction facilitates a next interaction and either one alone doesn't have the proper in and out amounts?

**Comment by nlordell** [comment]
> What if you have an interaction with one input amount and two output amounts or vice versa?

Good point. I guess these should be `inputs: Vec<TokenAmount>` and `outputs: Vec<TokenAmount>`.

> What if a first interaction facilitates a next interaction and either one alone doesn't have the proper in and out amounts?

Then, given ☝️, the first interaction be modelled as `{... input: [], output: [] ...}` and the second one containing a list of all input and output token amounts.

**Review by MartinquaXD — APPROVED** [review]

File: crates/shared/src/http_solver/model.rs [hunk 2]
```diff
@@ -121,14 +115,59 @@ pub struct ApprovalModel {
...
+    /// `AMM -> GPv2Settlement`
+    #[serde(default)]
+    pub outputs: Vec<TokenAmount>,
     pub exec_plan: Option<ExecutionPlanCoordinatesModel>,
 }
 
+/// Module to allow for backwards compatibility with the HTTP solver API.
+///
+/// Specifically, the HTTP solver API used to expect calldata as a JSON array of
+/// numbers in the range `[0, 256)`. This changed to allow `0x-` prefixed hex
```
Comment (MartinquaXD) [inline crates/shared/src/http_solver/model.rs:145]: super nit: `[0, 256)` seems like an odd notation given the fact that we are talking about integers.

**Review by MartinquaXD — APPROVED** [review]
Does still look good to me. 👍

File: crates/shared/src/http_solver/model.rs (hunk 2 above)
Comment (nlordell) [reply crates/shared/src/http_solver/model.rs]: Reworded.

**Pull request merged**